ProcessMaker/processmaker
# Integrate Office 365 SMTP Email Server

## Issue & Reproduction Steps

Ticket: [FOUR-6103](https://processmaker.atlassian.net/browse/FOUR-6103)

As a System Administrator, I want to enable an email server using the Office 365 SMTP so that my platform utilizes my preferred and familiar email service

This PR updates the Settings UI based on the [Email Server PRD](https://processmaker.atlassian.net/wiki/spaces/PM4/pages/2761064449/Email+Server+Options+-+Office365+Modern+Auth+and+Gmail+API) specifications as well as adds a new SMTP Authentication Method option allowing users to select either Standard, Google or Office 365 Authentication.

## Solution

- Adds the 'office365' option in the UI for the 'SMTP Authentication Method' Model
- Implements XOAUTH2 for Office 365 SMTP

## UI Updates

**Before**

![Screen Shot 2022-07-14 at 11 07 28 AM](https://user-images.githubusercontent.com/52755494/179053935-eff9820a-472f-45e7-b4a6-a206f9d77759.png)
![Screen Shot 2022-07-14 at 11 07 39 AM](https://user-images.githubusercontent.com/52755494/179053975-d779a316-99ea-4af3-b0ce-61a02b326c9d.png)

**After**

![Screen Shot 2022-08-10 at 9 09 35 AM](https://user-images.githubusercontent.com/52755494/184013948-0b2019de-1fbc-4cb9-a00b-1561a46df0a0.png)

![Screen Shot 2022-08-10 at 9 09 15 AM](https://user-images.githubusercontent.com/52755494/184013984-f9cdf669-dd37-4662-a1ad-323c92dcc437.png)


## How to Test


**Testing UI and Migration**

1. Review the [PRD](https://processmaker.atlassian.net/wiki/spaces/PM4/pages/2761064449/Email+Server+Options+-+Office365+Modern+Auth+and+Gmail+API)
2. Ensure you are on the `develop` branch for both Core and Send Email Connector
3. Ensure the following packages are installed - 
    - Auth Package
    - Send Email Connector  `develop`
    - Actions By Email
    - DocuSign Connector
4. Configure a few Email servers with and without SMTP settings
5. Update Core branch `feature/FOUR-6103B` 
6. Update Send Email Connector branch `feature/FOUR-6103B`
7. Reinstall Packages
4. Run `php artisan migrate`
8. Within the settings tabs ensure UI displays properly as specified in the PRD
   - Buttons are displayed properly
   - Existing Email Servers Configurations display properly
   - Add / Remove Email Servers works as expected
   - Configurations Update as expected. etc...


## Related Tickets & Packages
- Ticket: [FOUR-6103](https://processmaker.atlassian.net/browse/FOUR-6103)
- Ticket: [FOUR-6104](https://processmaker.atlassian.net/browse/FOUR-6104)
- [Email Server PRD](https://processmaker.atlassian.net/wiki/spaces/PM4/pages/2761064449/Email+Server+Options+-+Office365+Modern+Auth+and+Gmail+API)
- [Send Email PR - Google Integration](https://github.com/ProcessMaker/connector-send-email/pull/211) `feature/FOUR-6104B`

ci:connector-send-email:feature/FOUR-6103B

## Code Review Checklist
